Implement JSON Merge Patch format for PATCH requests

1zaman opened the pull request:
An OkHttp interceptor has been created that changes all PATCH requests with `application/json` MIME type to use `application/merge-patch+json` content type, according to the JSON Merge Patch format. The charset parameter is stripped out, as JSON is always encoded as Unicode, and it's specific encoding can be inferred. The manual Retrofit header annotations for this in the user profile API definition have been removed.

Since the HTTP headers are now being modified by interceptors, Retrofit logging has been disabled in favor of the OkHttp interceptor logger (logging will be removed in Retrofit 2 anyway). Moreover, the existing interceptor logger has been replaced with the one provided in the OkHttp `logging-interceptor` module, as of version 2.6. All the OkHttp modules have been upgraded from 2.5 to 2.7, and the unneeded dependency on `okhttp-urlconnection` has been removed.

The `LogUtil` class has also been changed to check `BuildConfig.DEBUG` instead of the "debuggable" flag before logging anything, in accordance with the new Gradle and Android Studio debug build paradigm, which also enables the compiler to remove the relevant code from release builds (as it's a constant flag).

https://openedx.atlassian.net/browse/MA-1695

bguertin commented:
:+1: I think we can also remove `@Headers("Content-Type: application/merge-patch+json")` from `UserService`

1zaman commented:
A few changes:

 - Removed manual Retrofit header annotations for this.
 - Content type is injected by OkHttp after application interceptors, so I created a wrapper for the `RequestBody` that returns `application/merge-patch+json` `MediaType`, and replaced the current body with the wrapper.
 - Retrofit logging has been disabled, and the `LoggingInterceptor` has been replaced with `HttpLoggingInterceptor`, which is provided in a separate module by OkHttp since version 2.6.
 - The `LogUtil` class has also been changed to check `BuildConfig.DEBUG` instead of the "debuggable" flag.

The details are included in the summary.

miankhalid commented:
:+1:
